include/seqan/basic/pair_packed.h

```diff
@@ -88,8 +88,20 @@ struct Pair<T1, T2, Pack>
 
     // Pair() = default; does not work on gcc4.9, it issues warnings if T1/T2
     // have no proper default constructor. >=gcc5.0 reports no warnings.
+#if defined(COMPILER_GCC) && (__GNUC__ <= 4)
     Pair() : i1(), i2() {};
+#else
+    Pair() = default;
+#endif
```

Comment:
These two mean different things. In one case the members will always be default-initialised, in the other case they are not. This has a potentially large impact on performance and/or correctness of code.
I see that this also changed in 8be002628c4047bcbdb10c6b823a63ea0a540a49 ...
Which of the behaviours is intended?

Reply:
Good catch, forgot to re-add the default initialization.

The original behaviour was
```c++
T1 i1;
T2 i2;
Pair() : i1(), i2() {};
```
After that I changed it to
```c++
T1 i1{};
T2 i2{};
Pair() = default;
```
That triggered warnings on gcc.

Thus changing it back to

```c++
T1 i1;
T2 i2;
Pair() : i1(), i2() {};
```

Now I thought that it might be better, if we ever drop a compiler, to make it explicit for newcomers.


```c++
T1 i1{};
T2 i2{};
#if ...
Pair() : i1(), i2() {};
#else
Pair() = default;
#endif
```

I might be wrong, but this code should not change the original behaviour, right? And should not have any performance impacts, right? 

Thank you!

Reply:
```c++
T1 i1{};
T2 i2{};
#if ...
Pair() : i1(), i2() {};
#else
Pair() = default;
#endif
```

is the same as 

```c++
T1 i1{};
T2 i2{};
Pair() = default;
```

for `#else` which is the normal case which will result in warning, or not? Did you try just using the member initialisers and removing all constructors so this is a simple aggregate type?

Reply:
I removed the other copy constructors, but some other code depended on them.

Reply:
> I removed the other copy constructors, but some other code depended on them.

You need to define all, or remove all. If all are removed, all are created by the compiler.

Reply:
I know, I did mean this copy-constructor:

```c++
     template <typename T1_, typename T2_, typename TSpec__>
     // TODO(holtgrew): explicit?
     Pair(Pair<T1_, T2_, TSpec__> const &_p) :
         i1(getValueI1(_p)), i2(getValueI2(_p))
     {}

Reply:
Here are some example errors from gcc-4.9:

```c++
seqan-src/include/seqan/basic/pair_packed.h:94:5: warning: missing initializer for member ‘seqan::Tuple<seqan::SimpleType<unsigned char, seqan::Dna_>, 7u, seqan::BitPacked<> >::i’ [-Wmissing-field-initializers]
     Pair() = default;
     ^

seqan-src/include/seqan/pipe/pipe_sampler.h:507:27: note: synthesized method ‘seqan::Pair<T1, T2, seqan::Tag<seqan::Pack_> >::Pair() [with T1 = seqan::Pair<unsigned char, unsigned int, seqan::Tag<seqan::Pack_> >; T2 = seqan::Tuple<seqan::SimpleType<unsigned char, seqan::Dna_>, 7u, seqan::BitPacked<> >]’ first required here 
             limits(_limits) {}
                           ^

seqan-src/include/seqan/basic/pair_packed.h:94:5: warning: missing initializer for member ‘seqan::Tuple<long unsigned int, 7u, seqan::Tag<seqan::Pack_> >::i’ [-Wmissing-field-initializers]
     Pair() = default;
     ^

seqan-src/include/seqan/pipe/pipe_sampler.h:118:25: note: synthesized method ‘seqan::Pair<T1, T2, seqan::Tag<seqan::Pack_> >::Pair() [with T1 = long unsigned int; T2 = seqan::Tuple<long unsigned int, 7u, seqan::Tag<seqan::Pack_> >]’ first required here 
             tmpRef(&tmp2) {}
                         ^

```

Reply:
Funny enough, if I use bracket `{}` initialization, I get the same warning. But then using `()` I get no error.

```c++
seqan-src/include/seqan/basic/pair_packed.h:92:23: warning: missing initializer for member ‘seqan::Tuple<seqan::SimpleType<unsigned char, seqan::Dna_>, 7u, seqan::BitPacked<> >::i’ [-Wmissing-field-initializers]
     Pair() : i1{}, i2{} {};
                       ^
```

I thought more about it, shouldn't the current implementation without member default initialization correspond to the old version. For me it looks like, that the `{}` initialization was buggy in gcc-4.9 and that they fixed it after that.

Reply:
but what about just removing all of the constructors? Let's discuss this on monday.

Reply:
Yep, you are right :)

BTW without any constructor I get:

```console
seqan-src/include/seqan/pipe/pipe_base.h:857:16: error: no matching function for call to ‘seqan::Pair<unsigned char, unsigned int, seqan::Tag<seqan::Pack_> >::Pair(int, int)’
         me.pos = TPair(0, 0);
                ^

seqan-src/include/seqan/index/index_skew7_multi.h:209:78: error: no matching function for call to ‘seqan::Pair<long unsigned int, long unsigned int, seqan::Tag<seqan::Pack_> >::Pair(long unsigned int, const long unsigned int&)’
             return TResult(off[seqOfs % 7][getValueI1(x_i1)] - seqOfs/7, x.i2);
                                                                              ^
```

Reply:
Yes, if you make it an aggregate type, you would need to do `TPair{0,0}`. But requiring that would be a breaking change so let's not do it.